Add preliminary support for MSG200

acascone opened the pull request:
I made some minor changes to support channels with the garage door opener, and have tested with the MSG100 and MSG200. I'm open to additional any additional changes, cleanup, or testing you'd like to see. This was just the minimal set of changes I needed to get it work.

The MSG100 uses channel 0, for the single door, but the MSG200 uses channels 1, 2, and 3. FWIW, setting channel 0 on the MSG200 to open or closed appears to control all doors.

donavanbecker commented:
@CocoaBob and @dylanfrankcom thoughts?

CocoaBob commented:
Yeah, the `channel` value wasn't variable for MSG100, so if MSG200 needs it, let's make it configurable. I can test it for MSG100 once it's merged (I remember we had a test branch or something similar?).

donavanbecker commented:
@CocoaBob yeah I would merge this to beta, let you test then we can release. I will do that.

donavanbecker commented:
@CocoaBob, @acascone latest beta has this. Can we test? and then I will merge this with master and release.

CocoaBob commented:
Just tested with `sudo npm -g install homebridge-plugins/homebridge-meross#beta`, and then Homebridge crashes. Here's the log, and I'm trying to figure it out.

```
systemd[1]: Started Node.js HomeKit Server.
homebridge[28500]: [11/18/2020, 3:58:37 PM] Loaded config.json with 2 accessories and 2 platforms.
homebridge[28500]: [11/18/2020, 3:58:37 PM] ---
homebridge[28500]: [11/18/2020, 3:58:38 PM] Loaded plugin: homebridge-config-ui-x@4.33.0
homebridge[28500]: [11/18/2020, 3:58:38 PM] Registering platform 'homebridge-config-ui-x.config'
homebridge[28500]: [11/18/2020, 3:58:38 PM] ---
homebridge[28500]: [11/18/2020, 3:58:38 PM] The plugin "homebridge-meross" requires Node version of >=14.15.1 which does not satisfy the current Node version of v12.19.0. You may need to upgrade your installation of Node.
homebridge[28500]: [11/18/2020, 3:58:38 PM] ====================
homebridge[28500]: [11/18/2020, 3:58:38 PM] ERROR LOADING PLUGIN homebridge-meross:
homebridge[28500]: [11/18/2020, 3:58:38 PM] Error: Cannot find module '/usr/lib/node_modules/homebridge-meross/dist/index.js'
homebridge[28500]: Require stack:
homebridge[28500]: - /usr/lib/node_modules/homebridge/lib/plugin.js
homebridge[28500]: - /usr/lib/node_modules/homebridge/lib/pluginManager.js
homebridge[28500]: - /usr/lib/node_modules/homebridge/lib/api.js
homebridge[28500]: - /usr/lib/node_modules/homebridge/lib/server.js
homebridge[28500]: - /usr/lib/node_modules/homebridge/lib/cli.js
homebridge[28500]: - /usr/lib/node_modules/homebridge/bin/homebridge
homebridge[28500]:     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:831:15)
homebridge[28500]:     at Function.Module._load (internal/modules/cjs/loader.js:687:27)
homebridge[28500]:     at Module.require (internal/modules/cjs/loader.js:903:19)
homebridge[28500]:     at require (internal/modules/cjs/helpers.js:74:18)
homebridge[28500]:     at Plugin.load (/usr/lib/node_modules/homebridge/src/plugin.ts:181:27)
homebridge[28500]:     at /usr/lib/node_modules/homebridge/src/pluginManager.ts:120:16
homebridge[28500]:     at Map.forEach (<anonymous>)
homebridge[28500]:     at PluginManager.initializeInstalledPlugins (/usr/lib/node_modules/homebridge/src/pluginManager.ts:118:18)
homebridge[28500]:     at Server.start (/usr/lib/node_modules/homebridge/src/server.ts:148:24)
homebridge[28500]:     at cli (/usr/lib/node_modules/homebridge/src/cli.ts:80:10)
homebridge[28500]: [11/18/2020, 3:58:38 PM] ====================
homebridge[28500]: [11/18/2020, 3:58:38 PM] Loaded plugin: homebridge-nest-cam@5.5.2
homebridge[28500]: [11/18/2020, 3:58:38 PM] Registering platform 'homebridge-nest-cam.Nest-cam'
homebridge[28500]: [11/18/2020, 3:58:38 PM] ---
homebridge[28500]: [11/18/2020, 3:58:38 PM] Loading 2 platforms...
homebridge[28500]: [11/18/2020, 3:58:38 PM] [Config] Initializing config platform...
homebridge[28500]: [11/18/2020, 3:58:38 PM] [Config] Spawning homebridge-config-ui-x with PID 28520
homebridge[28500]: [11/18/2020, 3:58:38 PM] [Nest-cam] Initializing Nest-cam platform...
homebridge[28500]: [11/18/2020, 3:58:38 PM] Loading 2 accessories...
homebridge[28500]: [11/18/2020, 3:58:38 PM] Error loading accessory requested in your config.json at position 1
homebridge[28500]: [11/18/2020, 3:58:38 PM] Error: The requested accessory 'Meross' was not registered by any plugin.
homebridge[28500]:     at PluginManager.getPluginForAccessory (/usr/lib/node_modules/homebridge/src/pluginManager.ts:198:15)
homebridge[28500]:     at /usr/lib/node_modules/homebridge/src/server.ts:345:37
homebridge[28500]:     at Array.forEach (<anonymous>)
homebridge[28500]:     at Server._loadAccessories (/usr/lib/node_modules/homebridge/src/server.ts:328:29)
homebridge[28500]:     at Server.start (/usr/lib/node_modules/homebridge/src/server.ts:154:12)
homebridge[28500]:     at cli (/usr/lib/node_modules/homebridge/src/cli.ts:80:10)
homebridge[28500]:     at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
homebridge[28500]:     at Module._compile (internal/modules/cjs/loader.js:1015:30)
homebridge[28500]:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
homebridge[28500]:     at Module.load (internal/modules/cjs/loader.js:879:32)
homebridge[28500]: [11/18/2020, 3:58:38 PM] Got SIGTERM, shutting down Homebridge...
systemd[1]: homebridge.service: Main process exited, code=exited, status=143/n/a
systemd[1]: homebridge.service: Failed with result 'exit-code'.
```

acascone commented:
@CocoaBob, I'm no Node expert, but it looks like the 3.3.2-beta bumped the Node version to 14 from 12. Perhaps that's the issue?

I was able to install the beta and it seems to be working, but I now see a new behavior that I hadn't seen when I was on my own fork. When I click to open a garage door, it says "Closing" even when it's supposed to be opening. The door does the correct thing, but this seems like a regression, and the issue isn't immediately obvious.

CocoaBob commented:
Sorry, my bad, everything's good, I should use `sudo npm -g install homebridge-meross@beta` to install the latest beta 😅

@acascone About the opening/closing status, actually this issue was there since my first implementation for MSG100 😅 I need more time to figure out a solution to guess/persist the actual opening/closing status.

CocoaBob commented:
@acascone There's an error on this line
```
for(var i = 0; i < response.payload.all.digest.garageDoor.length; i++)
```
It says
```
src/index.ts#L1275
Unexpected var, use let or const instead
```

Would you like to send another commit?

acascone commented:
I fixed the issues I introduced, but it looks like @donavanbecker had already taken care of that on the beta branch. It was a good exercise for me to get the linting working on my system tough.

CocoaBob commented:
Great, thanks! So let's wait for @donavanbecker to merge it, at least it works well on my side with MSG100 😊

donavanbecker commented:
I will merge #106 and release v3.4.0
